clustered/client/src/main/java/org/ehcache/clustered/client/internal/ConnectionSource.java

```diff
@@ -81,22 +90,58 @@ private static URI extractClusterUri(URI uri) {
 
   public static class ServerList extends ConnectionSource {
 
-    private final Iterable<InetSocketAddress> servers;
+    private final CopyOnWriteArraySet<InetSocketAddress> servers;
     private final String clusterTierManager;
 
     public ServerList(Iterable<InetSocketAddress> servers, String clusterTierManager) {
-      this.servers = cloneServers(Objects.requireNonNull(servers, "Servers cannot be null"));
+      this.servers = createServerList(servers);
       this.clusterTierManager = Objects.requireNonNull(clusterTierManager, "Cluster tier manager identifier cannot be null");
     }
 
+    private CopyOnWriteArraySet<InetSocketAddress> createServerList(Iterable<InetSocketAddress> servers) {
+      Objects.requireNonNull(servers, "Servers cannot be null");
+      CopyOnWriteArraySet<InetSocketAddress> serverList = new CopyOnWriteArraySet<>();
+      servers.forEach(server -> serverList.add(server));
+      return serverList;
+    }
+
     @Override
     public String getClusterTierManager() {
       return clusterTierManager;
     }
 
     @Override
     public LeasedConnection connect(Properties connectionProperties) throws ConnectionException {
-      return LeasedConnectionFactory.connect(servers, connectionProperties);
+      LeasedConnection connection = LeasedConnectionFactory.connect(servers, connectionProperties);
+      try {
+        EntityRef<DynamicTopologyEntity, Object, DynamicTopologyEntity.Settings> ref = connection.getEntityRef(DynamicTopologyEntity.class, 1, DynamicTopologyEntityConstants.ENTITY_NAME);
+        DynamicTopologyEntity dynamicTopologyEntity = ref.fetchEntity(null);
+        dynamicTopologyEntity.setListener(new DynamicTopologyEntity.Listener() {
+          @Override
+          public void onNodeRemoval(int stripeId, Node removedNode) {
+            servers.remove(removedNode.getNodeAddress());
+          }
+
+          @Override
+          public void onNodeAddition(int stripeId, Node addedNode) {
+            servers.add(addedNode.getNodeAddress());
+          }
+        });
+        return new LeasedConnection() {
+          @Override
+          public void close() throws IOException {
+            dynamicTopologyEntity.close();
```

Comment:
I don't think you can do this... if the connection is down I think this will block the connection from being closed.

Reply:
Planning to introduce async enity close for handling this.